torch/_dynamo/variables/nn_module.py

```diff
@@ -664,7 +670,28 @@ def call_method(
                     args,
                     kwargs,
                 )
+
             if id(method.__code__) in self._nn_module_method_ids():
                 unimplemented(f"UnspecializedNNModuleVariable missing {name}")
 
         return super().call_method(tx, name, args, kwargs)
+
+
+class FSDPNNModuleVariable(UnspecializedNNModuleVariable):
+    """
+    Tracing behavior: trace into submodules and treat them as Unspecialized, do not
+    register parameters to the top-level, treat them as function inputs.
+
+    Guards behavior: if 'skip_fsdp_guards', many guards that would be installed
+    by a vanilla UnspecializedNNMouduleVariable are simply dropped, on the basis
+    that a user wrapping their model in FSDP(model) is already opting into a
+    requirement to not modify internal model state, which would already break FSDP without
+    compilation.
+    """
+
+    def __init__(self, value, **kwargs):
+        super().__init__(value=value, **kwargs)
+        if torch._dynamo.config.skip_fsdp_guards:
+            self.source = FSDPNNModuleSource(self.source)
+        else:
+            self.source = NotNNModuleSource(self.source)
```

Comment:
Super for this class does:

`self.source = NotNNModuleSource(self.source)`

So I think you might end up with a `NotNNModuleSource` in the chain?

I think it would be clearer to move source writing into a method so that super always calls the lowest level of the hierarchy that overrides it, like https://github.com/pytorch/pytorch/pull/96905/files#diff-890f369c8fb7fd76d79bac7fe4f8dddf126916580d0a9ee6b64dad221ba3d46eR671